mir-ci/mir_ci/robot_libraries/Screencopy.py

```diff
@@ -0,0 +1,169 @@
+            )
+
+            return [
+                {
+                    "left": region.left,
+                    "top": region.top,
+                    "right": region.right,
+                    "bottom": region.bottom,
+                }
+                for region in regions
+            ]
+        except (RuntimeError, ValueError, ImageNotFoundError) as exc:
+            self._log_failed_match(screenshot, template)
+            raise ImageNotFoundError from exc
+
+    @keyword
+    async def wait_match(self, template: str, timeout: int = 10) -> List[dict]:
+        """
+        Grab screenshots and compare until there's a match with the provided
+        template.
+
+        :param template: path to an image file to be used as template
+        :param timeout: timeout in seconds
+        :return: list of matched regions
+        :raises ImageNotFoundError: if no match is found within the timeout
+        """
+        await self.connect()
+        regions = []
+        last_frame_count = 0
+        matched = False
+        screenshot = None
+        end_time = time.time() + float(timeout)
+        while time.time() < end_time:
+            if self.frame_count == last_frame_count:
+                await asyncio.sleep(0)
+            else:
+                last_frame_count = self.frame_count
+                try:
+                    screenshot = self.grab_screenshot()
+                    regions = self._rpa_images.find_template_in_image(
+                        screenshot,
+                        template,
+                        tolerance=self.TOLERANCE,
+                    )
+                    matched = True
+                    break
+                except (RuntimeError, ValueError, ImageNotFoundError):
+                    continue
+
+        if not matched:
+            self._log_failed_match(screenshot, template)
+            raise ImageNotFoundError
+
+        return [
+            {
+                "left": region.left,
+                "top": region.top,
+                "right": region.right,
+                "bottom": region.bottom,
+            }
+            for region in regions
+        ]
+
+    def grab_screenshot(self) -> Image.Image:
+        """Grab the latest captured frame from the display."""
+        assert self.shm_data is not None, "No SHM data available"
+
+        self.shm_data.seek(0)
+        data = self.shm_data.read()
+        size = (self.buffer_width, self.buffer_height)
+        stride = self.buffer_stride
+        image = Image.frombytes("RGBA", size, data, "raw", "RGBA", stride, -1)
+        b, g, r, a, *_ = image.split()
+        image = Image.merge("RGBA", (r, g, b, a))
+
+        return image
+
+    async def connect(self):
+        """Connect to the display."""
+        if not self.shm_data:
+            await super().connect()
+            # Wait for the first frame to become ready
+            while self.frame_count == 0:
+                await asyncio.sleep(0)
```

Comment:
Oh, this, shouldn't this be in `grab_screenshot` instead?

Reply:
Possibly, but it is a check that we don't need to do for every call of `grab_screenshot`. It's just a one-time check to make sure that we have at least one frame ready.